From 575986abefbb80b1fd2e7e7fc573935fc56c7bd8 Mon Sep 17 00:00:00 2001
From: Yang Chen <chenyang@cs.utah.edu>
Date: Thu, 26 Nov 2020 11:27:28 -0800
Subject: [PATCH] Handle using declarations declared with macros

In such cases, we should remove the macro expansions instead
---
 clang_delta/RemoveUnusedFunction.cpp                    | 8 +++++++-
 clang_delta/tests/remove-unused-function/macro_using.cc | 8 ++++++++
 2 files changed, 15 insertions(+), 1 deletion(-)
 create mode 100644 clang_delta/tests/remove-unused-function/macro_using.cc

diff --git a/clang_delta/RemoveUnusedFunction.cpp b/clang_delta/RemoveUnusedFunction.cpp
index 952f5a07..3c42dc5e 100644
--- a/clang_delta/RemoveUnusedFunction.cpp
+++ b/clang_delta/RemoveUnusedFunction.cpp
@@ -649,7 +649,13 @@ void RemoveUnusedFunction::removeOneFunctionDeclGroup(const FunctionDecl *FD)
     const FunctionDecl *ParentFD = UsingParentFDs[(*I).first];
     if (ParentFD && RemovedFDs.count(ParentFD->getCanonicalDecl()))
       continue;
-    RewriteHelper->removeDecl((*I).first);
+    const UsingDecl *UD = (*I).first;
+    SourceRange Range = UD->getSourceRange();
+    if (Range.getBegin().isMacroID()) {
+      TheRewriter.RemoveText(SrcManager->getExpansionRange(Range));
+    } else {
+      RewriteHelper->removeDecl((*I).first);
+    }
   }
 }
 
diff --git a/clang_delta/tests/remove-unused-function/macro_using.cc b/clang_delta/tests/remove-unused-function/macro_using.cc
new file mode 100644
index 00000000..3d264334
--- /dev/null
+++ b/clang_delta/tests/remove-unused-function/macro_using.cc
@@ -0,0 +1,8 @@
+// RUN: %clang_delta --transformation=remove-unused-function --counter=1 %s 2>&1 | %remove_lit_checks | FileCheck %s
+
+// CHECK: #define USING using ::foo;
+#define USING using ::foo;
+// CHECK-NOT: void foo
+void foo();
+// CHECK-NOT: USING 
+USING
